### ISSUE TITLE
Jupyterhub Proxy segfault on CentOS 7 when using SSL

### ISSUE THREAD

**Issue by zoltan-fedor** [jupyterhub/jupyterhub#833]
I was happily running JupyterHub on a CentOS 7 server for a while and suddenly after a server reboot it didn't start up anymore, so started investigating.

It turns out, that the JupyterHub Proxy kept restarting after poping a kernel segfault error:
`Oct 26 16:47:23 jupyterhub kernel: [  411.045850] traps: node[1507] general protection ip:2b1e33a1deb0 sp:7ffeb2b53150 error:0 in libcrypto.so.1.0.1e[2b1e33916000+1bd000]`

I am (was) running JupyterHub with SSL and it turns out everytime when it tries to use TLS the Proxy crashes with the above segfault.

Obviously this hasn't happened before, so I hoped that a downgrade of OpenSSL will solve it.
I looked at the yum.log and sure, there were recent openssl upgrades, so I downgraded, but none of the downgrades solved the issue. I have tried the following versions:

```
openssl-1.0.1e-51.el7_2.7 [current version]
openssl-1.0.1e-51.el7_2.5
openssl-1.0.1e-42.el7_1.9
```

As it seems this is not about the OpenSSL version, then likely it is about the kernel, so probably one of the recent kernel upgrades has caused it, but did not want to start downgrading kernels.
We are currently on kernel (latest) 3.10.0-327.36.3.el7.x86_64

It seems others have the same issue too (with NodeJS and JupyterHub too) - and so this might not be JupyterHub-specific, but maybe related to how it uses TLS:
https://lists.fedoraproject.org/archives/list/nodejs%40lists.fedoraproject.org/thread/F4G5CZI64IPIW5KNJK2B5M7QUX4X6JRA/

For now I have turned off SSL - I can do that, as we are using JupyterHub internally only.

**Comment by minrk** [comment]
It might be that you need to rebuild/reinstall whatever was linking openssl after upgrading openssl under it (perhaps nodejs itself?).

**Comment by ianabc** [comment]
I see something very similar to this on one of my machines. I haven't been able to reproduce it anywhere else, but on that one machine I've ditched the nodejs packages from EPEL and jumped to using. As a workaround I pulled a tarball of nodejs 4.x from https://nodejs.org/dist/latest-v4.x/, unpacked it in /opt, then symlinked node and npm into /usr/bin.

This is a really ugly workaround, but I couldn't afford much downtime for that machine and this avoided the segfaults.

FWIW the issue seems to be linking of nodejs and openssl in EPEL as @minrk suggests. I think I posted something to one of the nodejs lists, I'll try to dig that up and link it here.

**Comment by zoltan-fedor** [comment]
Thanks @ianabc 
It is interesting that you too had this happening on one of your machines but not on the others. I too has nodejs from EPEL, but really do not want to go down the route of having a locally compiled version on nodejs.
I think I am just going to stay on http for now (it is an internal service, so I can afford it for a while) and later switch to having an nginx in front of JupyterHub which will terminate SSL.

**Comment by ianabc** [comment]
I keep my eye on those EPEL packages and openssl hoping for an update which will fix it, I've tried to back out of the binaries from the nodejs tarball a couple of times, but no success so far! The machine it's on is busy which makes debugging the problem a real pain!

**Comment by zoltan-fedor** [comment]
Thanks.
I have the same problem - the machine is busy, so I don't really want to go on a wild goose chase recompiling stuff hoping that I will stumble on the right thing. Rather stay on http only for now, look for new EPEL packages and if they don't come or don't solve it in the next few weeks, then I will just go with putting NGinx in front with SSL termination.
In case if you come across with the right packages, please let us know.

**Comment by ianabc** [comment]
Will do! BTW the node packages include binaries for x86_64 so you don't need to recompile anything, the main problems is remembering that I've put them in there when it comes time to try the some new EPEL packages :-)

**Comment by willingc** [comment]
@ianabc Thanks for clarifying the issue and solution until EPEL has linked nodejs and OpenSSL suitably. @zoltan-fedor Please keep us in the loop if you see the EPEL update come out.

I'm going to label this issue "Out of Scope" as it depends on the EPEL update and set the milestone to review the status at the end of November. @minrk

**Comment by zoltan-fedor** [comment]
I've just tested it with the latest nodejs package from EPEL (nodejs-6.9.1-4.el7.x86_64) which came out a few days ago.
Unfortunately the issue is still there.

Will try again when newer packages come out and let you know when it finally gets fixed in the EPEL nodejs package.

**Comment by willingc** [comment]
Thanks @zoltan-fedor. Please keep us posted with EPEL. cc/@minrk

**Comment by willingc** [comment]
Status note: Checked EPEL and nodejs on https://lists.fedoraproject.org to see if there has been a more recent release than the one referenced above. Not yet.

**Comment by willingc** [comment]
Status update: The latest [nodejs package on EPEL is (nodejs-6.9.4-2.el7)](https://lists.fedoraproject.org/archives/list/epel-package-announce@lists.fedoraproject.org/message/2FQ6LQ373RURKRFBDDKD5DKBRDFKAKGF/) came out on 2-7-17.

If this resolves the issue, please let us know. Thanks.

**Comment by zoltan-fedor** [comment]
Thanks, just tested nodejs package on EPEL is (nodejs-6.9.4-2.el7) and unfortunately it does NOT resolve the issue. It is still the same.

**Comment by willingc** [comment]
@zoltan-fedor Thanks for testing. 😞 that it didn't resolve the issue. 

I have this on my radar to check monthly.

**Comment by toddpfaff** [comment]
I tested this today with centos 7 epel nodejs-6.10.1-2.el7.x86_64 and the problem still exists.

This issue is also discussed here: https://lists.fedoraproject.org/archives/list/nodejs@lists.fedoraproject.org/thread/F4G5CZI64IPIW5KNJK2B5M7QUX4X6JRA/#LPU5XR75NAWBU3WMUEFF56PNOVCLGYZN

@zoltan-fedor How did you make out with an nginx front-end to jupyterhub?

**Comment by zoltan-fedor** [comment]
@toddpfaff , the nginx front-end works just fine. Since then I have deployed multiple Jupyterhub instances with NGinx front terminating SSL. Works very stable, no problems whatsoever.

**Comment by vroomanj** [comment]
I've encountered this same issue. I'm using Let's Encrypt and this problem occurs when using the fullchain. If you use just the cert (without the CA) this problem goes away. That leaves JupyterHub open to a man-in-the-middle attack. Running nginx in front and using a firewall makes it quite unlikely but it would be nice to either be able to use the fullchain or include the CA.

I believe the following 2 errors are related to this:
> [E 2017-06-08 16:58:33.541 JupyterHub ioloop:638] Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x7f9949b21158>, <tornado.concurrent.Future object at 0x7f9949af57f0>)
>     Traceback (most recent call last):
>       File "/usr/lib64/python3.4/site-packages/tornado/ioloop.py", line 605, in _run_callback
>         ret = callback()
>       File "/usr/lib64/python3.4/site-packages/tornado/stack_context.py", line 277, in null_wrapper
>         return fn(*args, **kwargs)
>       File "/usr/lib64/python3.4/site-packages/tornado/ioloop.py", line 626, in _discard_future_result
>         future.result()
>       File "/usr/lib/python3.4/site-packages/jupyterhub/app.py", line 1217, in check_proxy
>         yield self.proxy.add_all_services(self.services)
>       File "/usr/lib/python3.4/site-packages/jupyterhub/orm.py", line 222, in add_all_services
>         service = service_dict[orm_service.name]
>     TypeError: list indices must be integers, not str

> [E 2017-06-08 18:40:01.341 JupyterHub ioloop:638] Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x7fdb92073d08>, <tornado.concurrent.Future object at 0x7fdb93757940>)
>     Traceback (most recent call last):
>       File "/usr/lib64/python3.4/site-packages/tornado/ioloop.py", line 605, in _run_callback
>         ret = callback()
>       File "/usr/lib64/python3.4/site-packages/tornado/stack_context.py", line 277, in null_wrapper
>         return fn(*args, **kwargs)
>       File "/usr/lib64/python3.4/site-packages/tornado/ioloop.py", line 626, in _discard_future_result
>         future.result()
>       File "/usr/lib/python3.4/site-packages/jupyterhub/app.py", line 1214, in check_proxy
>         yield self.start_proxy()
>       File "/usr/lib/python3.4/site-packages/jupyterhub/app.py", line 1197, in start_proxy
>         _check()
>       File "/usr/lib/python3.4/site-packages/jupyterhub/app.py", line 1193, in _check
>         raise e
>     RuntimeError: Proxy failed to start with exit code -11

CC: @willingc @minrk

**Comment by vroomanj** [comment]
```
# cat /etc/centos-release
CentOS Linux release 7.3.1611 (Core)

# python3 --version
Python 3.4.5

# jupyterhub --version
0.7.2

# node --version
v6.10.3

# configurable-http-proxy --version
2.0.2

# openssl version
OpenSSL 1.0.1e-fips 11 Feb 2013
```

**Comment by minrk** [comment]
@vroomanj what's the error output from the proxy subprocess when this happens? Running CHP with letsencrypt fullchain is what I normally do, so something funky is definitely going on, either with the nodejs on CentOS, openssl, or the link between the two.

One option to get around this could be to run CHP in a docker container instead of directly on the host:

    docker run \
    --name=jupyterhub-proxy \
    --restart=always \
    -e CONFIGPROXY_API_TOKEN=`secret` \
    -p 443:8000 \
    -v /etc/letsencrypt:/etc/letsencrypt \
    jupyterhub/configurable-http-proxy:2.0.3 \
        --default-target=http://<jupyterhub-ip>:8081 \
        --ssl-key=/etc/letsencrypt/live/example.org/privkey.pem \
        --ssl-cert=/etc/letsencrypt/live/example.org/fullchain.pem

**Comment by willingc** [comment]
Has anyone tested this with the latest [EPEL 7 nodejs release](https://lists.fedoraproject.org/archives/list/epel-package-announce@lists.fedoraproject.org/message/LWEPWLRE3TAEG6QVMAC4BJRWZQDRQGQN/)?

**Comment by richardtin** [comment]
Hi all,

I think my segfault problem may be the same with this issue. (I use a socket.io server with SSL)
```
[1124473.388005] node[8348]: segfault at 1 ip 00007f76b83f1acc sp 00007fff878a02a0 error 4 in libcrypto.so.1.0.2k[7f76b82cd000+234000]
```

* System Info
```
CentOS Linux release 7.3.1611 (Core)
node v6.9.4
libcrypto.so.1.0.2k
```

Finally, I solved this issue by upgrading nodejs to **v6.12.3**

**Comment by willingc** [comment]
Closing this issue. Thank you @richardtin for the update :tada:

[EPEL 7 nodejs release v6.12.3](https://lists.fedoraproject.org/archives/list/epel-package-announce@lists.fedoraproject.org/message/E2LDEULCUHXIJC3HN5BRQNNF6IBKB56G/)
